tests/components/motionblinds_ble/test_cover.py

```diff
@@ -0,0 +1,113 @@
+"""Tests for Motionblinds BLE covers."""
+
+from typing import Any
+from unittest.mock import patch
+
+from motionblindsble.const import MotionBlindType, MotionRunningType
+from motionblindsble.device import MotionDevice
+import pytest
+
+from homeassistant.components.cover import (
+    ATTR_POSITION,
+    ATTR_TILT_POSITION,
+    DOMAIN as COVER_DOMAIN,
+    SERVICE_CLOSE_COVER,
+    SERVICE_CLOSE_COVER_TILT,
+    SERVICE_OPEN_COVER,
+    SERVICE_OPEN_COVER_TILT,
+    SERVICE_SET_COVER_POSITION,
+    SERVICE_SET_COVER_TILT_POSITION,
+    SERVICE_STOP_COVER,
+    SERVICE_STOP_COVER_TILT,
+    STATE_CLOSED,
+    STATE_CLOSING,
+    STATE_OPEN,
+    STATE_OPENING,
+)
+from homeassistant.components.motionblinds_ble.const import DOMAIN
+from homeassistant.const import ATTR_ENTITY_ID, Platform
+from homeassistant.core import HomeAssistant
+
+from . import setup_platform
+
+
+@pytest.mark.parametrize(
+    ("service", "method", "kwargs"),
+    [
+        (SERVICE_OPEN_COVER, "open", {}),
+        (SERVICE_CLOSE_COVER, "close", {}),
+        (SERVICE_OPEN_COVER_TILT, "open_tilt", {}),
+        (SERVICE_CLOSE_COVER_TILT, "close_tilt", {}),
+        (SERVICE_SET_COVER_POSITION, "position", {ATTR_POSITION: 5}),
+        (SERVICE_SET_COVER_TILT_POSITION, "tilt", {ATTR_TILT_POSITION: 10}),
+        (SERVICE_STOP_COVER, "stop", {}),
+        (SERVICE_STOP_COVER_TILT, "stop", {}),
+    ],
+)
+async def test_cover_service(
+    hass: HomeAssistant, service: str, method: str, kwargs: dict[str, Any]
+) -> None:
+    """Test cover service."""
+
+    _, name = await setup_platform(
+        hass, [Platform.COVER], blind_type=MotionBlindType.VENETIAN
+    )
+
+    with patch(
+        f"homeassistant.components.motionblinds_ble.MotionDevice.{method}"
+    ) as func:
+        await hass.services.async_call(
+            COVER_DOMAIN,
+            service,
+            {ATTR_ENTITY_ID: f"cover.{name}", **kwargs},
+            blocking=True,
+        )
+        func.assert_called_once()
+
+    await hass.async_block_till_done()
+
+
+@pytest.mark.parametrize(
+    ("running_type", "state"),
+    [
+        (None, "unknown"),
+        (MotionRunningType.STILL, "unknown"),
+        (MotionRunningType.OPENING, STATE_OPENING),
+        (MotionRunningType.CLOSING, STATE_CLOSING),
+    ],
+)
+async def test_cover_update_running(
+    hass: HomeAssistant, running_type: str | None, state: str
+) -> None:
+    """Test updating running status."""
+
+    config_entry, name = await setup_platform(
+        hass, [Platform.COVER], blind_type=MotionBlindType.VENETIAN
+    )
+
+    device: MotionDevice = hass.data[DOMAIN][config_entry.entry_id]
```

Comment:
Don't touch internals in tests

Reply:
I can't access the MotionDevice?

Reply:
Nope, mock everything and let the mocks play out the things you want to happen

Reply:
How should I do that then? Each entity registers callbacks on the MotionDevice when it is added. The MotionDevice is created in the async_setup_entry and then set in hass.data: `hass.data.setdefault(DOMAIN, {})[entry.entry_id] = device`. How do I make sure the device is the Mock in that case and therefore each entity registers a callback on the Mock?

Reply:
Checkout `knocki` there I eventually fetch the callback function added to the client and call that to update states

Reply:
Do you know why using the mock_motion_device results in the unique_id changing from "cover.motionblind_cccc" (I guess that is the entry_id) to "cover.motionblinds_ble_cc_cc_cc_cc_cc_cc"

```
@pytest.mark.parametrize(
    ("running_type", "state"),
    [
        (None, "unknown"),
        (MotionRunningType.STILL, "unknown"),
        (MotionRunningType.OPENING, STATE_OPENING),
        (MotionRunningType.CLOSING, STATE_CLOSING),
    ],
)
async def test_cover_update_running(
    mock_config_entry: MockConfigEntry,
    mock_motion_device: Mock,  # Adding this line changes the unique_id
    hass: HomeAssistant,
    running_type: str | None,
    state: str,
) -> None:
    """Test updating running status."""

    name = await setup_integration(hass, mock_config_entry)

    async_update_running = mock_motion_device.register_running_callback.call_args[0][0]

    async_update_running(running_type)
    assert hass.states.get("cover.motionblinds_ble_cc_cc_cc_cc_cc_cc").state == state
    assert hass.states.get("cover.{name}").state == state
    # ^^ The last assertion suddenly doesn't work anymore since the unique_id changed when using mock_motion_device
```

Reply:
Cover takes the device name and the device name is set to `device.display_name`. Is `display_name` properly mocked? Otherwise I think it will either follow the config entry title or the unique_id if `display_name` would suddently return nothing

Reply:
At this point I have no idea what is happening to the names. I have hardcoded some of the names in the code (see latest commits) to get 100% coverage but I am fully confused as to how and why these names are generated like this, and the logic required to get this working without hardcoding it:
- button.mock_title_connect
- button.disconnect
- cover.motionblinds_ble_cc_cc_cc_cc_cc_cc
- cover.motionblinds_cccc

Here are some of the states with all of these unique_ids:
[<state button.mock_title_connect=unknown; friendly_name=Connect @ 2024-07-13T10:07:24.430987-07:00>, <state button.disconnect=unknown; friendly_name=Disconnect @ 2024-07-13T10:07:24.432163-07:00>, <state button.favorite=unknown; friendly_name=Favorite @ 2024-07-13T10:07:24.432977-07:00>, <state cover.motionblinds_ble_cc_cc_cc_cc_cc_cc=unknown; device_class=shade, friendly_name=, supported_features=15 @ 2024-07-13T10:07:24.440620-07:00>, <state select.speed=unknown; options=['1', '2', '3'], friendly_name=Speed @ 2024-07-13T10:07:24.444602-07:00>, <state sensor.connection_status=disconnected; options=['connected', 'connecting', 'disconnected', 'disconnecting'], device_class=enum, friendly_name=Connection status @ 2024-07-13T10:07:24.449583-07:00>, <state sensor.signal_strength=unknown; unit_of_measurement=dBm, device_class=signal_strength, friendly_name=Signal strength @ 2024-07-13T10:07:24.451120-07:00>, <state sensor.battery=unknown; state_class=measurement, unit_of_measurement=%, device_class=battery, friendly_name=Battery @ 2024-07-13T10:07:24.451763-07:00>]

Reply:
You were right! display_name was indeed set to an empty string causing this odd behaviour. Fixed it now